### PR TITLE
fix: enforce worksheet resource budgets

### LINKED ISSUE #17: [P1] Enforce a real export iteration budget for sparse far-edge cells
## Context

`sheetToJson` (`src/api/json.ts`), `sheetToCsv` (`src/api/csv.ts`) and `sheetToHtml` (`src/api/html.ts`) iterate from `range.s` to `range.e` of the **declared** `!ref`/dimension, not over the cells that actually exist.

## Problem (verified in code)

A worksheet can legally declare `!ref = "A1:XFD1048576"` (16,384 × 1,048,576 ≈ 17 billion cell positions) while containing only a handful of real cells. Each exporter then runs a dense double loop over the declared range:

- `sheetToJson` allocates one output object per row and probes every column → tens of billions of `getCell` calls and an enormous output array.
- `sheetToCsv` / `sheetToHtml` likewise build a string per row/column.

A ~2 KB crafted file can wedge the process for minutes or OOM it — a cheap denial-of-service on any server that reads a sheet and converts it.

## Proposed solution

- For sparse sheets, clamp the iteration to the actual occupied range (intersect the declared range with `Object.keys`-derived bounds, or with the dense `!data` length).
- Honor `sheetRows` consistently across all three exporters (it currently influences parsing/JSON but the export loops still trust `e.r`).
- Optionally expose a `maxCells` budget in the opts and throw when exceeded.
- Add a test with a tiny file declaring a giant `!ref` and assert it completes quickly / throws rather than hanging.

## Why it matters

Pairs with the ZIP-bomb hardening: both are required before the library is safe to point at untrusted uploads.

_Filed from a code-quality/maturity audit, 2026-06-12._

### LINKED ISSUE #86: [P1] Apply parsing budgets to derived worksheet ranges and HTML spans
## Priority and impact

**P1** — Small inputs can request disproportionate CPU/memory despite configured limits. A shared expansion budget must cover every parser path.

## Verified evidence

Audited on 2026-09-08 against `main` (`c3c1e30`) plus the isolated Unicode decoding fix (`69e8b93`). This is independent of the decoding bug fixed in #84.

The worksheet cell counter covers explicit `<c>` elements, but hyperlink ranges create additional cells without charging that counter. With a configured budget of one cell and one row, a bounded 20-by-10 hyperlink range produced 200 cells. Column definitions similarly expanded to 200 entries without a dimension budget.

The plain-text HTML path drops the read options altogether: a small two-row table with a 100-column span still expanded with `sheetRows: 1`, `maxWorksheetRows: 1`, and `maxWorksheetCells: 1`.

These loops scale with numeric dimensions from input rather than the number of XML/HTML elements. The earlier raw XML limits in #25 do not cover this generated work. No oversized resource-exhaustion payload was run.

Code: [src/xlsx/worksheet.ts:58](https://github.com/sebastian-software/xlsx-format/blob/69e8b9372b6ff27ab1fea29e8a0cd63313e5ead4/src/xlsx/worksheet.ts#L58), [src/xlsx/worksheet.ts:113](https://github.com/sebastian-software/xlsx-format/blob/69e8b9372b6ff27ab1fea29e8a0cd63313e5ead4/src/xlsx/worksheet.ts#L113), [src/api/html.ts:265](https://github.com/sebastian-software/xlsx-format/blob/69e8b9372b6ff27ab1fea29e8a0cd63313e5ead4/src/api/html.ts#L265), [src/read.ts:91](https://github.com/sebastian-software/xlsx-format/blob/69e8b9372b6ff27ab1fea29e8a0cd63313e5ead4/src/read.ts#L91).

## Acceptance criteria

- [ ] Validate row, column, and span dimensions before expansion.
- [ ] Charge hyperlink-created cells and column/span expansion against documented budgets and reject before allocating.
- [ ] Forward applicable limits and sheetRows through read() into text/HTML parsing.
- [ ] Add bounded over-budget tests for sparse/dense XLSX and HTML; invalid/negative/non-finite dimensions must reject predictably.

## Delivery estimate

Effort: **M**, including regression tests. Fix risk: **MEDIUM**. Confidence: **HIGH** (direct code trace and/or bounded synthetic reproduction). No production data was used.

### PR DESCRIPTION
**Pull request by swernerx**
Small worksheet elements could request disproportionate work through hyperlink ranges, column definitions, HTML spans or merges, and far occupied ranges. Exporters could still iterate a huge rectangle after the existing occupied-range clamp.

This change makes `maxWorksheetCells` a documented cumulative per-sheet work budget. XLSX, CSV, and HTML imports charge explicit and derived work before allocation, validate physical worksheet coordinates, and apply `sheetRows` consistently to cells, hyperlinks, HTML rows, and `rowspan="0"` groups. Plain-text `read()` forwards the configured CSV and HTML limits.

CSV, JSON, HTML, formula-list, and XLSX exports clamp to occupied data and then enforce a 1,000,000-position default. XLSX output writes the effective clamped dimension, and sparse row, column, shared-string, and HTML-merge paths visit occupied metadata without array-length or position-by-metadata multiplication. Valid numeric JSON starts beyond the data continue to return an empty result.

Validation:

- `oxfmt --check src/`
- `eslint src/`
- `tsc --noEmit`
- `vitest run --coverage` (1,158 tests; 93.71% lines, 85.50% branches)
- `pnpm run package:check` (build, publint, attw, export smoke)
- `pnpm --filter docs build`

Fixes #86
Fixes #17

### CODE REVIEW

**Review by swernerx — COMMENTED** [review]
Review is on hold for two concrete corrections. (1) The new range preflight makes sheetToJson/sheetToArray(arrayToSheet([["header"]]), { range: 1 }) throw INVALID_ARGUMENT. A valid numeric start beyond available rows must retain the previous empty-result behavior; invalid numeric options should still reject. (2) Dense SST resolution still traverses data.length and each row.length in worksheet.ts. A worksheet with one far-column cell per row charges only one cell per row but performs up to 16,384 slot checks per row afterward. Iterate actual occupied row/cell keys so the new import work boundary is not defeated by dense sparse coordinates. Both findings have been returned to the implementing agent for focused regressions and integrated verification.

File: src/api/csv.ts [hunk 1]
```diff
@@ -185,14 +195,52 @@ export function sheetToTxt(sheet: WorkSheet, opts?: Sheet2CSVOpts): string {
...
 	const rows: any[][] = [];
 	let row: any[] = [];
 	let i = 0;
 	const len = text.length;
 	let afterSeparator = false;
+	let rowActive = false;
+	let rowCount = 0;
+	const maxRows = worksheetOptionLimit(opts.maxWorksheetRows, XLSX_MAX_ROWS, "maxWorksheetRows");
+	const sheetRows = worksheetOptionLimit(opts.sheetRows, 0, "sheetRows");
+	const budget = new WorksheetCellBudget(opts.maxWorksheetCells, DEFAULT_MAX_IMPORT_CELLS);
```
[the remark this answers is not in the document]
Comment (swernerx) [reply src/api/csv.ts:208]: Fixed in 39f8ece: plain-text `read()` now forwards options to `htmlToSheet`, matching CSV. The public regression covers `sheetRows`, `maxWorksheetRows`, and `maxWorksheetCells` through `read(..., { type: "string" })`.

File: src/xlsx/worksheet.ts [hunk 2]
```diff
@@ -296,6 +326,9 @@ function parseSheetData(
 				}
 			}
 			C -= 1; // Convert to 0-based
+			if (!Number.isSafeInteger(C) || C < 0 || C >= XLSX_MAX_COLUMNS) {
+				throw new XlsxError("MALFORMED", "Invalid worksheet cell column: value exceeds XLSX worksheet bounds");
+			}
```
[the remark this answers is not in the document]
Comment (swernerx) [reply src/xlsx/worksheet.ts]: Fixed in 63bb83e: parsed cell references now validate both physical row and column bounds and require the cell row suffix to match its containing `<row>`. Sparse and dense regressions cover an out-of-bounds suffix and a mismatched row.

File: src/xlsx/worksheet.ts [hunk 3]
```diff
@@ -806,11 +855,15 @@ export function writeWorksheetXml(ws: WorkSheet, opts: any, _idx: number, _rels:
 	lines.push("<sheetData>");
 
 	const dense = ws["!data"] != null;
-	const range = safeDecodeRange(ref);
+	const range = ws["!ref"] ? clampLargeExportRange(ws, safeDecodeRange(ref), budget) : null;
```
[the remark this answers is not in the document]
Comment (swernerx) [reply src/xlsx/worksheet.ts]: Fixed in 63bb83e: XLSX output computes the effective occupied range before writing `<dimension>` and uses that same range for `<sheetData>`. The regression writes a full-sheet declared range with one nearby occupied cell and verifies both the XML dimension and read-back `!ref` are clamped.

**Review by swernerx — COMMENTED** [review]
Final review at 39f8ecec3afecbcf38dc8c6c3ab5ea33e047ed05: no remaining blocking findings. The previously reported empty JSON range and dense shared-string scans are fixed. Independent public-API probes cover cumulative export budgets, physical bounds, sparse/dense hyperlink clipping, HTML read option forwarding, bounded row groups, and indexed HTML merges (100 merge entry reads for 400 output positions). A separate reviewer reproduced the corrected limits and passed 116 focused tests plus TypeScript. The author reports all 1,158 tests, formatting, lint, TypeScript, package validation, export smoke tests and documentation build passing. Source and regression review are complete; merge remains conditional on all CI results including the external coverage checks.
